docs/manual/src/udl/enumerations.md

```diff
@@ -17,4 +18,25 @@ enum Animal {
 };
 ```
 
-Note that enumerations with associated data are not yet supported.
+Enumerations with associated data require a different syntax,
+due to the limitations of using WebIDL as the basis for UniFFI's interface language.
+An enum like this in Rust:
+
+```rust
+enum IpAddr {
+  V4 {q1: u8, q2: u8, q3: u8, q4: u8},
+  V6 {addr: string},
+}
+```
+
+Can be exposed in the UDL file with:
+
+```idl
+[TaggedUnion]
+interface IpAddr {
+  V4(u8 q1, u8 q2, u8 q3, u8 q4);
+  V6(string addr);
+};
+```
```

Comment:
This is the proposed UDL syntax. I think it looks nice, but it's a pretty serious abuse of WebIDL syntax.  It parses, but e.g. the `V4` line actually parses as an anonymous special method returning a datatype named `V4` rather than as a field or method named `V4`. I personally have no problem with us re-interpreting that as a tagged union, but it seems gross so YMMV. I haven't been able to come up with anything better though :-/

I had hoped to be able to do something clever with the existing WebIDL `enum` declaration, along the lines of:

```
[TaggedUnion]
enum IpAddr {
  [Fields(u8)] "V4",
  [Fields(u8, string)] "V6",
}
```

But I couldn't come up with something that would parse. In particular, attributes don't seem to be supported on the variants of an enum.

Reply:
I love the syntax you've arrived at.

This and #379 reminds me we never quite resolved the question of writing our own parser. Soon is coming the time to reconsider [ADR-0001][1]!

[1]: https://github.com/mozilla/uniffi-rs/blob/600882a0a36191ad4565bda62a22841984d777ae/docs/adr/0001-mvp-webidl.md

Reply:
> Soon is coming the time to reconsider ADR-0001!

Yes! I have many half-baked thoughts on this. What do you think is a good way to structure that kind of discussion, should we e.g. open an open-ended "discuss" issue where we can kick some ideas around?

Reply:
+1 to what James said, although I'm slightly more skeptical we will find time to rewrite the parser until that's our final option :)

Reply:
I agree with Mark here: I expect we'll only seriously consider doing the work once we've exhausted.

That'd be the wrong time for a discussion, so yes, let's start a doc/issue and iterate.

Reply:
> That'd be the wrong time for a discussion, so yes, let's start a doc/issue and iterate.

Sounds good 👍 

(FWIW, I was pretty close to declaring "WebIDL has no good syntax for this" before I discovered the anonymous-special-method-returning-a-named-datatype hackery you see here, so I won't be surprised if we run out of runway with WebIDL sooner than we expect)

Reply:
@rfk So will any WebIDL we generate (not just the internal UDL) for Firefox have a way of expressing this stuff?

Reply:
I filed a "discuss" issue here: https://github.com/mozilla/uniffi-rs/issues/385